src/main/java/uk/nhs/digital/nhsconnect/lab/results/model/edifact/segmentgroup/InvolvedParty.java

```diff
@@ -0,0 +1,56 @@
+package uk.nhs.digital.nhsconnect.lab.results.model.edifact.segmentgroup;
+
+import static java.util.stream.Collectors.toList;
+
+import java.util.List;
+import java.util.Optional;
+import lombok.Getter;
+import lombok.NonNull;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.MessageRecipientNameAndAddress;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.PerformingOrganisationNameAndAddress;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.RequesterNameAndAddress;
+import uk.nhs.digital.nhsconnect.lab.results.model.edifact.ServiceProvider;
+
+/**
+ * Represents information about involved healthcare parties.
+ * <p>
+ * Segment group 1: {@code S01-NAD-ADR-COM-RFF-SEQ-SPR}
+ * <p>
+ * Parents: {@link uk.nhs.digital.nhsconnect.lab.results.model.edifact.Message Message}
+ */
+@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+public class InvolvedParty extends SegmentGroup {
+    public static final String INDICATOR = "S01";
+
+    // NAD+SLA?
+    @Getter(lazy = true)
+    private final Optional<PerformingOrganisationNameAndAddress> performingOrganisationNameAndAddress =
+        extractOptionalSegment(PerformingOrganisationNameAndAddress.KEY_QUALIFIER)
+            .map(PerformingOrganisationNameAndAddress::fromString);
+
+    // NAD+PO?
+    @Getter(lazy = true)
+    private final Optional<RequesterNameAndAddress> requesterNameAndAddress =
+        extractOptionalSegment(RequesterNameAndAddress.KEY_QUALIFIER)
+            .map(RequesterNameAndAddress::fromString);
+
+    // NAD+MR?
+    @Getter(lazy = true)
+    private final Optional<MessageRecipientNameAndAddress> messageRecipientNameAndAddress =
+        extractOptionalSegment(MessageRecipientNameAndAddress.KEY_QUALIFIER)
+            .map(MessageRecipientNameAndAddress::fromString);
+
+    // SPR
```

Comment:
Are we missing a getter for `RFF`?

Reply:
Yeah, that segment doesn't exist in our codebase. Adding them was out of scope really for this story, but since I added some of the others I'll add this too.